std/algorithm/iteration.d

```diff
@@ -5001,61 +5001,107 @@ private struct SplitterResult(alias isTerminator, Range)
 }
 
 /++
-Lazily splits the string `s` into words, using whitespace as the delimiter.
+Lazily splits the character-based range `s` into words, using whitespace as the
+delimiter.
 
-This function is string specific and, contrary to
+This function is character-range specific and, contrary to
 `splitter!(std.uni.isWhite)`, runs of whitespace will be merged together
 (no empty tokens will be produced).
 
 Params:
-    s = The string to be split.
+    s = The character-based range to be split. Must be a string, or a
+    random-access range of character types.
 
 Returns:
     An $(REF_ALTTEXT input range, isInputRange, std,range,primitives) of slices of
-    the original string split by whitespace.
+    the original range split by whitespace.
  +/
-auto splitter(C)(C[] s)
-if (isSomeChar!C)
+auto splitter(Range)(Range s)
+if (isSomeString!Range ||
+    isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range &&
+    !isConvertibleToString!Range &&
+    isSomeChar!(ElementEncodingType!Range))
 {
     import std.algorithm.searching : find;
     static struct Result
     {
     private:
         import core.exception : RangeError;
-        C[] _s;
+        Range _s;
         size_t _frontLength;
 
-        void getFirst() pure @safe
+        void getFirst()
         {
             import std.uni : isWhite;
 
-            auto r = find!(isWhite)(_s);
-            _frontLength = _s.length - r.length;
+            static if (is(ElementType!Range == dchar) ||
+                       is(ElementType!Range == const(dchar)) ||
+                       is(ElementType!Range == immutable(dchar)))
+            {
+                // dchar range, we can just use find.
```

Comment:
Hm... is there not a danger of new whitespace characters being added to the spec at a future date?

Reply:
Note that stripLeft and stripRight only take a shortcut for ascii, otherwise it decodes.

Reply:
For new whitespace characters to be a problem they would have to be outside the basic multilingual plane. No whitespace has been added to Unicode since 2002, and no whitespace characters outside the "General Punctuation" / "Supplemental Punctuation" blocks have been added since 1999.

Reply:
>Note that `stripLeft` and `stripRight` only take a shortcut for ascii, otherwise it decodes.

The "shortcut" in `stripLeft` is a pessimization for auto-decoding types. It uses `.front` so for `string` and `wstring` it performs decoding once and if the result is non-ASCII it performs decoding a second time. That's among the things I fix in PR #6727.

Reply:
Hm.. I get confused with all the template constraints. I thought `!isConvertibleToString` would disqualify normal string arrays. I actually don't know what it means when something passes or fails that check!

So I thought that overload was only for non-auto-decoding strings, which would actually not make it a pessimization.

I will review your other PR.

However, in this case, we are just using `find` on auto-decoding strings (i.e. element type is `dchar`), which should do the right thing.



> For new whitespace characters to be a problem they would have to be outside the basic multilingual plane. No whitespace has been added to Unicode since 2002, and no whitespace characters outside the "General Punctuation" / "Supplemental Punctuation" blocks have been added since 1999.

OK, good evidence that we can take the risk. I'll update for `wchar`.